File: assets/android-engine/P2y.java
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2026  Nija (bubu12) and contributors
//
// P2y —— 普通话书面文 → 粤语口语(供 TTS 朗读用)。原生 Java,零依赖。
//
// ⚠️ 本文件是 Python 版 `/usr/local/bin/p2y` 的【逐条机械复刻】,不是重新设计。
//    判据:同一批语料下,Java 输出与 Python 输出【逐字节相同】。
//    (Python 版已过 157 单元用例 + CER 实测 23.7%→4.3%;复刻一致即继承那些结论)
//
// ⚠️ 六步处理顺序(顺序敏感,别调):
//   ⓪ 总开关(VSAY_P2Y=0 / VSAY_CANTO_P2Y=0 ⇒ 原样返回)
//   ① 半角标点归全角 —— 只在【紧邻 CJK 汉字】时转
//      为什么不能无脑全局转:会把 `127.0.0.1:8790` 变成 `127.0.0.1：8790`,
//      把 `a,b` 变成 `a，b`;而 DSH 正文里夹着端口号/路径/代码片段。
//   ② 进行体「在+V」→「喺度+V」(正则)
//   ③ 数字+元/块 → 蚊(正则,不参与词表)
//   ④ 【带 \x00 哨兵】的单遍最长匹配
//      🔑 哨兵的用意:把"文本末尾"变成普通字符,这样"句末 了→喇"就能写成
//         普通表条目("了\x00" → "喇\x00")⇒ 与"长词优先"共用同一套机制。
//         ⚠️ 我第一版漏了哨兵 ⇒ 句末语气词规则【全部失效】。
//   ⑤ 收尾压掉重复助词(单遍匹配后基本不会出现,留作兜底)
//
// 用法:
//   P2y p = P2y.load(inputStream);   // assets/p2y-rules.tsv
//   String yue = p.convert("谢了,这个功能已经完成了。");

package canto;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class P2y {

    /** 句末哨兵(中英文里绝不可能出现 ⇒ 零误伤) */
    private static final char SENTINEL = '\u0000';

    /** 半角 → 全角(只这五种) */
    private static final char[][] PUNCT = {
        {',', '，'}, {'?', '？'}, {'!', '！'}, {';', '；'}, {':', '：'}
    };
    private static final Map<Character, Character> PUNCT_MAP = new HashMap<>();
    static {
        for (char[] p : PUNCT) PUNCT_MAP.put(p[0], p[1]);
    }

    /** 进行体:「在」后面紧跟这些动词之一 ⇒ 换成「喺度」 */
    private static final Pattern IN_PROGRESS =
        Pattern.compile("在(?=[等看聽听說说講讲想找做寫写食吃喝跑忙玩睡])");
    /** 数字 + 元 / 块 ⇒ 蚊 */
    private static final Pattern DIGIT_MONEY  = Pattern.compile("(\\d)\\s*元");
    private static final Pattern DIGIT_MONEY2 = Pattern.compile("(\\d)\\s*[块塊]");
    /** 收尾压重复助词 */
    private static final Pattern DUP_GE  = Pattern.compile("(嘅){2,}");
    private static final Pattern DUP_ZO  = Pattern.compile("(咗){2,}");
    private static final Pattern DUP_HAI = Pattern.compile("(係){2,}");

    private final Map<String, String> rules;   // src -> dst(已含带哨兵的条目)
    private final int maxSrcLen;

    private P2y(Map<String, String> rules, int maxSrcLen) {
        this.rules = rules;
        this.maxSrcLen = maxSrcLen;
    }

    /** 从 TSV 流加载(一行一条:`源词\t目标词`;源词里可能含 \u0000 哨兵) */
    public static P2y load(InputStream in) throws IOException {
        Map<String, String> m = new HashMap<>(4096);
        int max = 1;
        BufferedReader br = new BufferedReader(new InputStreamReader(in, "UTF-8"));
        try {
            String line;
            while ((line = br.readLine()) != null) {
                if (line.isEmpty()) continue;
                int tab = line.indexOf('\t');
                if (tab <= 0) continue;
                String src = line.substring(0, tab);
                String dst = line.substring(tab + 1);
                if (!m.containsKey(src)) m.put(src, dst);   // 先出现者赢(PROTECT 在 LEX 前)
                if (src.length() > max) max = src.length();
            }
        } finally {
            br.close();
        }
        return new P2y(m, max);
    }

    // ────────────────────────── ① 半角标点归全角 ──────────────────────────

    /** 是否 CJK 汉字(与 Python 版 _is_cjk 同判据) */
    private static boolean isCjk(char c) {
        return (c >= 0x4E00 && c <= 0x9FFF)     // CJK 统一表意
            || (c >= 0x3400 && c <= 0x4DBF)     // 扩展 A
            || (c >= 0xF900 && c <= 0xFAFF)     // 兼容表意
            || (c >= 0x3040 && c <= 0x30FF);    // 日文假名(粤语文本里也可能混)
    }

    /** 只在标点【左边或右边紧邻 CJK】时才转全角 */
    static String normPunct(String t) {
        boolean any = false;
        for (int i = 0; i < t.length(); i++) {
            if (PUNCT_MAP.containsKey(t.charAt(i))) { any = true; break; }
        }
        if (!any) return t;
        char[] out = t.toCharArray();
        int n = t.length();
        for (int i = 0; i < n; i++) {
            Character to = PUNCT_MAP.get(t.charAt(i));
            if (to == null) continue;
            char prev = i > 0 ? t.charAt(i - 1) : 0;
            char next = i + 1 < n ? t.charAt(i + 1) : 0;
            if (isCjk(prev) || isCjk(next)) out[i] = to;
        }
        return new String(out);
    }

    // ────────────────────── ④ 带哨兵的单遍最长匹配 ──────────────────────

    private String onePass(String text) {
        int n = text.length();
        StringBuilder sb = new StringBuilder(n + 16);
        int i = 0;
        while (i < n) {
            int remain = n - i;
            int tryMax = remain < maxSrcLen ? remain : maxSrcLen;
            boolean matched = false;
            for (int len = tryMax; len >= 1; len--) {
                String cand = text.substring(i, i + len);
                String hit = rules.get(cand);
                if (hit != null) {
                    sb.append(hit);
                    i += len;
                    matched = true;
                    break;
                }
            }
            if (!matched) {
                sb.append(text.charAt(i));
                i++;
            }
        }
        return sb.toString();
    }

    // ────────────────────────────── 主流程 ──────────────────────────────

    /** 六步转换。enabled=false 时原样返回(对应 VSAY_P2Y=0) */
    public String convert(String text, boolean enabled) {
        if (text == null || text.isEmpty()) return text;
        if (!enabled) return text;

        String t = text;
        t = normPunct(t);                                        // ①
        t = IN_PROGRESS.matcher(t).replaceAll("喺度");            // ②
        t = DIGIT_MONEY.matcher(t).replaceAll("$1蚊");            // ③
        t = DIGIT_MONEY2.matcher(t).replaceAll("$1蚊");
        t = onePass(t + SENTINEL);                               // ④ 带哨兵
        t = t.replace(String.valueOf(SENTINEL), "");
        t = DUP_GE.matcher(t).replaceAll("嘅");                   // ⑤ 兜底
        t = DUP_ZO.matcher(t).replaceAll("咗");
        t = DUP_HAI.matcher(t).replaceAll("係");
        return t;
    }

    public String convert(String text) { return convert(text, true); }

    public int ruleCount() { return rules.size(); }

    /** 调试用:看某一步的输出 */
    public String debugStep(String text, int step) {
        String t = text;
        if (step >= 1) t = normPunct(t);
        if (step >= 2) t = IN_PROGRESS.matcher(t).replaceAll("喺度");
        if (step >= 3) { t = DIGIT_MONEY.matcher(t).replaceAll("$1蚊");
                         t = DIGIT_MONEY2.matcher(t).replaceAll("$1蚊"); }
        if (step >= 4) { t = onePass(t + SENTINEL).replace(String.valueOf(SENTINEL), ""); }
        if (step >= 5) { t = DUP_GE.matcher(t).replaceAll("嘅");
                         t = DUP_ZO.matcher(t).replaceAll("咗");
                         t = DUP_HAI.matcher(t).replaceAll("係"); }
        return t;
    }
}

File: assets/android-p2y/P2y.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2026  Nija (bubu12) and contributors
//
// P2y —— 普通话书面文 → 粤语口语(供 TTS 朗读用)
//
// ⚠️ 设计要点(为什么这么写):
//   1) 【不把 1907 条规则写进代码】—— 规则表是【数据】,随包走(assets/p2y-rules.tsv),
//      这样:① Kotlin 侧只有薄实现,易维护 ② 规则可独立更新,不用重编 App
//   2) 【单遍最长匹配】—— 输出字符【永不再扫描】。上游 Python 版 v1 用 replace 链,
//      会级联污染(`但是太貴了` → `但係係太貴喇`);v2 改单遍最长匹配才修好。我们复刻 v2 语义。
//   3) 规则表【已按长度降序排好】(上游 `p2y --list` 就是这么排的)⇒ 顺序扫描即可,
//      第一个命中的就是最长匹配,不需要额外的 trie。
//   4) 【保护表已并入同一张表】—— 上游把 PROTECT(身份保留,如"目的"→"目的")和
//      LEX(真正转换)合成一张表,靠最长优先天然生效 ⇒ 我们不用分开处理。
//
// ⚠️ 与 Python 版的一致性:本实现是【机械复刻】,不是"重新设计"。
//    验证方式:对同一批语料,Android 输出应与 `p2y` 输出【逐字节相同】。

package com.fcitx5sensevoice.canto

import android.content.Context
import java.io.BufferedReader
import java.io.InputStreamReader

class P2y private constructor(
    /** 规则表:src → dst;已按 src 长度降序 + 字典序排好 */
    private val rules: Array<Pair<String, String>>,
    /** 最长 src 的长度(用于剪枝:剩余文本不足这么长就不用试了) */
    private val maxSrcLen: Int,
) {

    companion object {
        const val ASSET_NAME = "p2y-rules.tsv"

        /** 从 assets 加载(TSV:源词<TAB>目标词,一行一条,已排序) */
        fun fromAssets(ctx: Context): P2y {
            val list = ArrayList<Pair<String, String>>(2048)
            var maxLen = 1
            ctx.assets.open(ASSET_NAME).use { ins ->
                BufferedReader(InputStreamReader(ins, Charsets.UTF_8)).use { br ->
                    br.forEachLine { line ->
                        if (line.isEmpty()) return@forEachLine
                        val tab = line.indexOf('\t')
                        if (tab <= 0) return@forEachLine
                        val src = line.substring(0, tab)
                        val dst = line.substring(tab + 1)
                        list.add(src to dst)
                        if (src.length > maxLen) maxLen = src.length
                    }
                }
            }
            return P2y(list.toTypedArray(), maxLen)
        }

        /** 测试用:直接给规则表 */
        fun fromRules(rules: List<Pair<String, String>>): P2y =
            P2y(rules.toTypedArray(), rules.maxOfOrNull { it.first.length } ?: 1)
    }

    /**
     * 转换。**单遍最长匹配** —— 输出的字符不再参与后续匹配(防级联污染)。
     *
     * 算法:
     *   i 从 0 起;在位置 i,尝试长度从 min(maxSrcLen, 剩余) 递减到 1 的 src
     *   ⇒ 第一个命中就用它,把 dst 追加到输出,i 前进 src.length
     *   ⇒ 都不命中就把 text[i] 原样追加,i 前进 1
     */
    fun convert(text: String): String {
        if (text.isEmpty()) return text
        val sb = StringBuilder(text.length + 16)
        var i = 0
        val n = text.length
        while (i < n) {
            val remain = n - i
            val tryMax = if (remain < maxSrcLen) remain else maxSrcLen
            var matched = false
            var len = tryMax
            while (len >= 1) {
                val cand = text.substring(i, i + len)
                val hit = lookup(cand)
                if (hit != null) {
                    sb.append(hit)
                    i += len
                    matched = true
                    break
                }
                len--
            }
            if (!matched) {
                sb.append(text[i])
                i++
            }
        }
        return sb.toString()
    }

    /**
     * 查表。规则表已排序,但为了 O(1) 仍建索引。
     * ⚠️ 规则表里同一 src 可能出现多次(PROTECT + LEX),【先出现的赢】(与上游 set() 后的行为一致
     *    —— 上游是 set(PROTECT + LEX),PROTECT 在前 ⇒ 保护条目优先)。
     */
    private val index: HashMap<String, String> by lazy {
        HashMap<String, String>(rules.size * 2).apply {
            for ((s, d) in rules) putIfAbsent(s, d)   // 先出现者赢
        }
    }

    private fun lookup(s: String): String? = index[s]
}
